### DISCUSSION TITLE
Playlist Weights

### DISCUSSION THREAD

**Discussion by wppronto** [AzuraCast/AzuraCast#7032] [Q&A]
This is similar to my post about rotation issues, but this is more specific to weights.

I am testing with these playlists:

1970s: Weight 5
1980s: Weight 6
1990s: Weight 4

1990s
1980s
1970s
1980s
1990s
1970s
1970s
1990s
1990s
1970s
1990s
1970s
1970s
1970s
1980s
1990s

So that was a pretty decent hour, mixed up nicely. I didn't like having three songs in a row from the 1970s playlist, especially since I checked the box to only play one item from this playlist.

1970s: 7 (weight: 5)
1980s: 3 (weight: 6)
1990s: 6 (weight: 4)

However, the totals above don't make sense. The 1980s list should have had the most songs, but it had the fewest.

The next hour, with the same settings was a bit of a train wreck.

1970s
1990s
1970s
1980s
1970s
1980s
1980s
1980s
1980s
1980s
1980s
1980s
1980s
1980s
1980s
1970s

1970s: 4 (weight: 5)
1980s: 11 (weight: 6)
1990s: 1 (weight: 4)

The totals here made more sense, though I don't know why the 1980s playlist had almost there times as many songs played as the 1970s list, even though there is only one number of difference in playlist weight.

And, 10 of the 11 1980s songs were played in a row!

What am I doing wrong?

**Comment by gAlleb** [comment]
Hi, @wppronto! 

It's important to consider the fact that `default mechanism` which Liquidsoap inside Azuracast uses when you specify weights in your playlist configuration is not `rotate` but `random`: `radio = random(id="standard_playlists", weights=[1,2,3], [playlist1,playlist2,playlist3])`

This means that when you choose weights for your playlists - the won't play specifically like you think but in more `random` manner. The weight argument says how many tracks of each source should be taken in average. The selection is randomized however and it might happen that it doesn't follow your instructions perfectly. 

Plus and what's most important - on top of all of that we have Azuracast's own complex nextsong API which does the calculation in it's own way: checks for next songs depending on initial parameters and passes them back to Liquidsoap. Therefore some "tricks" and "magic" can happen along this way.

Alternatively, if you use `advanced playlist option` and manually set weights with `rotate` (not `random`) operator in your `edit liquidsoap configuration`: `radio = rotate(weights=[1,2,3,4,], [playlist1,playlist2,playlist3, radio])` - Then you will disable `Azuracast's nextsong API` for that playlists and Liquidsoap will try to comply to your weights rotation. However it might be not accurate sometimes too. It's just how it is for now.

**Comment by wppronto** [comment]
Hello [gAlleb](https://github.com/gAlleb)!

Thanks. While I know it's all approximate, my concern is that the weight settings just don't seem to be honored at all, or barely so. And the "Only play one song from this playlist at the specified time" option just seems to be ignored. So it's definitely hard to program when the results differ so widely from what you entered in terms of the weights.

I did try Liquidsoap rotations, and I was able to get the results I wanted. Of course I could even get granular by specifying each individual song slot (one from this playlist, one from that, etc.). But I hate relying on that because any tweaks will require reloading the configuration. I know that's not supposed to boot listeners, but it seems to in my testing, or am I wrong on that?

But your answer is very helpful because it explains that the default mechanism is random, whereas custom Liquidsoap code allows me to choose rotation instead, and dial it in pretty well if I can live with using custom Liquidsoap code.

**Reply by gAlleb** [reply]
If you choose `reload configuration` It will restart liquidsoap, but won’t restart Icecast. Depending on timeout your listeners will be or won’t be disconnected. 

That’s why `next song api` is so convenient: you do not need to restart liquidsoap each time you make changes to your playlists (except some) or make new ones, because song rotation is handled by the “outside” script. But sometimes it lacks more complex custom rules handling.

**Comment by wppronto** [comment]
A pseudo-related question: If I'm using Advanced Playlists and custom Liquidsoap code to handle rotation, will playlist schedules still work?

For example, let's say I have a Liquidsoap config that rotates through my 1970s, 1980s and 1990s playlists. But, in November and December, I want to add 1 Christmas song per rotation. Can I create the Christmas playlist and schedule it for November and December, then include it in the Liquidsoap rotation year-round (but it will be ignored those 10 months out of the year)? Or does making it Advanced override the scheduling?

Thanks,

Mark

**Reply by gAlleb** [reply]
Overrides. Playlist schedules won't work since only you now define how a playlist should behave.

But you can manually make it work. For example like this:

1. Create a `christmas_playlist` and make it advanced.

2. Define a function which would act like time interval using unix timestamps: from November 15th to December 31st (you can convert time to timestamps via online converter for example).

```ruby

def christmas_date_range() =
    # 2024-11-15 01:00:00
    range_start = 1731632400.
    # 2024-12-31 23:59:59
    range_end = 1735678799.
    current_time = time()
    result = (range_start <= current_time and current_time <= range_end)
    result
end

```

3. Your rotating playlists. Say you are rotating them 1-2-3, so 6 tracks = full rotation:

```
radio = rotate(weights=[1,2,3], [playlist1,playlist2,playlist3])
```

4. Then we add  a `switch` with our function inside of it, which activates your `christmas playlist` on `christmas date range`.

```
radio = switch(track_sensitive=true, [
({ christmas_date_range() }, rotate(weights=[1,6], [christmas_playlist, radio])), ({true}, radio)
])

```

6. This would rotate 3 of your playlists all year long but from November to December an additional `christmas_playlist` will kick in with 1 track and 6 tracks from your rotating playlists.

**Comment by wppronto** [comment]
Wow! Thank you so much!

**Comment by wppronto** [comment]
For regular playlists, it would be great if there was a setting to "Play no more than X songs from this playlist in a row." Without that, and with the behavior I'm seeing, sometimes there are 8-9 songs in a row from the same list, whereas I'm trying to do a variety format.

**Reply by wppronto** [reply]
Another thought, the way the random stuff works, the playlist weights don't work as the interface states. It says something like "playlists with higher weights will be played more often", but I'm not seeing that happen.

**Reply by gAlleb** [reply]
There is also a possibility that if you use 'enable_autocue_metadata' - it may confuse Azuracast sometimes due to computations. Also it could confuse Azuracast if you have short jingles cause there could be some problems with them and they take part in "rotation" too like ordinary playlists.. Try switching it off - maybe you'll see a difference.

You can also look up `Autocue2` protocol by @Moonbase59. It has proven itself faster in computations with some additional features => https://github.com/Moonbase59/autocue. I personally use it with simple crossfade with no fade_in and minimum fade_out to get the segue effect.

You could also see the debug in "Broadcasting" - "Upcoming Song Queue" - "logs" to see if there are any errors in playback queuing. Also restarting and clearing cache might help too.
<img width="500" alt="Снимок экрана 2024-03-31 в 15 25 11" src="https://github.com/AzuraCast/AzuraCast/assets/144447266/0d0cd164-75ca-48eb-b47c-d121818c6e32">

**Reply by Moonbase59** [reply]
I agree, I have also seen multiple jingles playing back-to-back when the upcoming jingle after a song was shorter than cross duration. This seems to be a Liquidsoap issue, and the devs are currently working on it.

Looks like kind of emergency "don’t interrupt playout" feature. When LS sees the buffer running empty, it apparently pulls the next track immediately.

**Comment by wppronto** [comment]
Thank you. In this test case I'm not running autocue, though looking forward to testing the latest developments there!

**Comment by wppronto** [comment]
Again, I think the uneven playlist weights could be resolved with a "Don't play more than X songs in a row from this playlist" option.

**Reply by gAlleb** [reply]
> I think the uneven playlist weights could be resolved with a "Don't play more than X songs in a row from this playlist" option.

If we change context of what you are saying - we could the same way say:

`I think the world hunger could be resolved with a "Give them all food" option.` ;)

We are talking here about somewhat complex `nextsong API` working via PHP. As I see it: it does a pretty good job. I have 4 general playlists with 2,2,2,2 weights and they work just fine. For me "general rotation playlists" are not so important and they're usually used to fill up space between scheduled programs and LiveDJ streaming. And there we can try more exquisit radio programming.

If you want a more precise option - you can use advanced playlists with a switch and manual coding or use "Manual AutoDJ" thus disabling `nextsong API` but I must say that getting tracks via the API from the database is a wonderful thing)

**Comment by wppronto** [comment]
Thanks, I see your point. However, I still think this is a valid concern, and that my suggestion would help. Your playlists work with the same weight each, but weights are also meant for situations like mine, to help certain playlists play more often. But it just isn't working that way. My suggestion would help.

And I do also recommend that AzuraCast add a "Give everyone food" option. ;-)
